### PR TITLE
Implement caching mechanism for v2/stacks API

### PR DESCRIPTION
**Pull request by abdermendz**
Description:
Issue: LMCROSSITXSADEPLOY-2284
This PR provides optimization caching mechanism for the v2/stacks API.
Reason: Reducing the unnecessary and repetitive calls to v2/stacks API.

### CODE REVIEW

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientFactory.java [hunk 1]
```diff
@@ -1,12 +1,12 @@
 package com.sap.cloudfoundry.client.facade.rest;
 
-import java.net.URL;
-import java.time.Duration;
-import java.util.Collections;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
-
+import com.sap.cloudfoundry.client.facade.CloudCredentials;
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientFactory.java:3]: Why are the imports rearranged?

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 2]
```diff
@@ -1,27 +1,71 @@
...
-import java.util.Optional;
-import java.util.Set;
-import java.util.UUID;
-import java.util.concurrent.TimeUnit;
-import java.util.function.Function;
-import java.util.function.IntFunction;
-import java.util.function.Supplier;
-import java.util.stream.Collectors;
-
+import com.sap.cloudfoundry.client.facade.CloudCredentials;
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:3]: Same here

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 3]
```diff
@@ -256,12 +258,13 @@ protected CloudControllerRestClientImpl() {
     }
 
     public CloudControllerRestClientImpl(URL controllerUrl, CloudCredentials credentials, WebClient webClient, OAuthClient oAuthClient,
-                                         CloudFoundryClient delegate) {
-        this(controllerUrl, credentials, webClient, oAuthClient, delegate, null, null);
+                                         CloudFoundryClient delegate, CloudStackCache cloudStackCache) {
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:261]: This object already has too many parameters to its constructor. Adding more isn't great

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 4]
```diff
@@ -717,15 +721,39 @@ public InstancesInfo getApplicationInstances(CloudApplication application) {
...
+
+    private CloudApplication addStackToCloudApplication(CloudApplication application) {
+        UUID cloudStackId = application.getStaging()
+                                       .getStackId();
+        CloudStack cloudStack = retrieveCloudStack(cloudStackId, () -> ImmutableRawCloudStack.of(getStackResource(cloudStackId).block())
+                                                                                             .derive());
+        return copyCloudApplicationWithStacks(application, cloudStack).derive();
+    }
+
+    private CloudStack retrieveCloudStack(UUID rawCloudStackId, Supplier<CloudStack> cloudStack) {
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:741]: The name of the CF entity is just `Stack`. We don't need to call it by our wrapper object's name `CloudStack`.
Another thing, the name of the second parameter isn't descriptive of what it is.

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 5]
```diff
@@ -717,15 +721,39 @@ public InstancesInfo getApplicationInstances(CloudApplication application) {
...
+                                                                     .withStackName(cloudStack.getName()));
     }
 
     @Override
     public List<CloudApplication> getApplicationsByMetadataLabelSelector(String labelSelector) {
         Map<String, Metadata> applicationsMetadata = getApplicationsMetadataByLabelSelector(labelSelector);
-        List<CloudApplication> cloudApplications = fetchListWithAuxiliaryContent(() -> getApplicationResourcesByNamesInBatches(applicationsMetadata.keySet()),
-                                                                                 this::zipWithAuxiliaryApplicationContent);
-        return addMetadata(cloudApplications, applicationsMetadata);
+        return addMetadata(addStacksToCloudApplications(fetchListWithAuxiliaryContent(() -> getApplicationResourcesByNamesInBatches(applicationsMetadata.keySet()),
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:754]: Why is this inlined? 3 function calls is hard to read. Split them into calling and assignment.

File: src/main/java/com/sap/cloudfoundry/client/facade/util/CloudStackCache.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package com.sap.cloudfoundry.client.facade.util;
+
+import com.sap.cloudfoundry.client.facade.CloudOperationException;
+import com.sap.cloudfoundry.client.facade.domain.CloudStack;
+import org.springframework.util.ConcurrentReferenceHashMap;
+
+import java.util.UUID;
+import java.util.function.Supplier;
+
+public class CloudStackCache {
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/util/CloudStackCache.java:10]: Maybe `StacksCache` would be better

File: src/main/java/com/sap/cloudfoundry/client/facade/util/CloudStackCache.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    private final ConcurrentReferenceHashMap<UUID, CloudStack> cloudStacks = new ConcurrentReferenceHashMap<>();
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/util/CloudStackCache.java:12]: `cachedStacks`?

File: src/main/java/com/sap/cloudfoundry/client/facade/util/Messages.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.sap.cloudfoundry.client.facade.util;
+
+public class Messages {
+
+    // ERROR messages
+    public static final String COULD_NOT_RETRIEVE_CLOUD_STACK = "Could not retrieve cloud stack by id.";
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/util/Messages.java:6]: A "cloud stack" doesn't exist. Just "stack"s do. Prefer not to use our custom implementation names for CF entities.

File: src/test/java/com/sap/cloudfoundry/client/facade/CloudStackIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+
+class CloudStackIntegrationTest extends CloudControllerClientIntegrationTest {
+
+    private static final String DEFAULT_STACK_NAME = "cflinuxfs3";
+    private static final String INCORRECT_STACK_NAME = "Non-existent-stack-name";
+    private static final CloudRouteSummary DEFAULT_CLOUD_ROUTE_SUMMARY = ImmutableCloudRouteSummary.builder()
```
Comment (radito3) [inline src/test/java/com/sap/cloudfoundry/client/facade/CloudStackIntegrationTest.java:31]: `default_route` is sufficient

File: src/test/java/com/sap/cloudfoundry/client/facade/CloudStackIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+    @BeforeAll
+    static void createDefaultDomain() {
+        client.addDomain(DEFAULT_DOMAIN);
+    }
+
+    @AfterAll
+    static void deleteDefaultDomain() {
+        List<CloudRoute> routes = client.getRoutes(DEFAULT_DOMAIN);
+        routes.forEach(route -> client.deleteRoute(route.getHost(), DEFAULT_DOMAIN, null));
```
Comment (radito3) [inline src/test/java/com/sap/cloudfoundry/client/facade/CloudStackIntegrationTest.java:44]: "Use lambdas judiciously."<sup>Joshua Bloch, Effective Java</sup> This use of `forEach` is harder to read than an ordinary range for loop.

File: src/test/java/com/sap/cloudfoundry/client/facade/IntegrationTestConstants.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package com.sap.cloudfoundry.client.facade;
+
+public class IntegrationTestConstants {
```
Comment (radito3) [inline src/test/java/com/sap/cloudfoundry/client/facade/IntegrationTestConstants.java:3]: nice

File: src/test/java/com/sap/cloudfoundry/client/facade/util/CloudStackCacheTest.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.mockito.Mockito;
+
+import java.util.UUID;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Comment (radito3) [inline src/test/java/com/sap/cloudfoundry/client/facade/util/CloudStackCacheTest.java:13]: What does this do?

File: src/main/java/com/sap/cloudfoundry/client/facade/util/CloudStackCache.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+    public void removeStackFromCache(UUID cloudStackId) {
+        cloudStacks.remove(cloudStackId);
+    }
+
+    public boolean containsCloudStack(UUID cloudStackId) {
+        return this.cloudStacks.containsKey(cloudStackId);
+    }
+
+    public int getCurrentCacheSize() {
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/util/CloudStackCache.java:30]: As far as I see, these 3 methods are only used in tests. Do we really need them?

File: src/main/java/com/sap/cloudfoundry/client/facade/domain/Staging.java [hunk 14]
```diff
@@ -61,7 +61,10 @@
      * @return the stack to use when staging the application, or null to use the default stack
      */
     @Nullable
-    String getStack();
+    String getStackName();
+
+    @Nullable
+    UUID getStackId();
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/domain/Staging.java:67]: Why is this needed here?

File: src/test/java/com/sap/cloudfoundry/client/facade/util/CloudStackCacheTest.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+class CloudStackCacheTest {
```
Comment (radito3) [inline src/test/java/com/sap/cloudfoundry/client/facade/util/CloudStackCacheTest.java:14]: Since the cache is just basically a wrapper around a hashtable, is this test class _really_ needed?

**Review by radito3 — CHANGES_REQUESTED** [review]
Prefer not to use implementation-specific names such as `CloudStack` when naming variables or as part of method names. Use the CF entity name `Stack`.
The approach here seems a bit overkill and unnecessary. Wouldn't it be simpler to just call the cache as a part of the `.zipWhen(this::getApplicationStackResource)` call? As in, modifying only the `getAppStack` method. The modification to the `Staging` object seems redundant.

File: src/main/java/com/sap/cloudfoundry/client/facade/domain/Staging.java (hunk 14 above)
Comment (abdermendz) [reply src/main/java/com/sap/cloudfoundry/client/facade/domain/Staging.java:67]: Because it's the only way to retrieve the id of a stack in order to use it as a key for the StackCache. At least the only logical way since we retrieve the stack's name from here.

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudApplication.java [hunk 16]
```diff
@@ -1,9 +1,12 @@
 package com.sap.cloudfoundry.client.facade.adapters;
 
 import java.util.List;
+import java.util.Optional;
 import java.util.Set;
+import java.util.UUID;
 import java.util.stream.Collectors;
 
+import com.sap.cloudfoundry.client.facade.Nullable;
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudApplication.java:9]: Unused imports Optional, UUID & Nullable

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 17]
```diff
@@ -1690,12 +1689,15 @@ private CloudApplication findApplication(UUID guid) {
...
-                                                     .map(tuple -> ImmutableRawCloudApplication.builder()
-                                                                                               .application(application)
-                                                                                               .summary(tuple.getT1())
-                                                                                               .stack(ImmutableRawCloudStack.of(tuple.getT2()))
-                                                                                               .space(target)
-                                                                                               .build());
+                                                     .doOnSuccess(tuple -> stacksCache.setStack(UUID.fromString(tuple.getT1()
+                                                                                                                     .getStackId()),
+                                                                                                tuple.getT2()))
+                                                     .map(summary -> ImmutableRawCloudApplication.builder()
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:1695]: Revert the name of this variable back to `tuple`

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 18]
```diff
@@ -1712,6 +1714,11 @@ private CloudApplication findApplication(UUID guid) {
 
     private Mono<? extends Resource<StackEntity>> getApplicationStackResource(SummaryApplicationResponse summary) {
         UUID stackGuid = UUID.fromString(summary.getStackId());
+        if (stacksCache.containsStack(stackGuid)) {
+            System.out.println("====CACHE IS USED ====");
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:1718]: https://www.youtube.com/watch?v=xoMgnJDXd3k

File: src/main/java/com/sap/cloudfoundry/client/facade/util/StacksCache.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    public Resource<StackEntity> getCloudStack(UUID stackId) {
+        return cachedStacks.get(stackId);
+    }
+
+    public boolean containsStack(UUID stackId) {
+        return cachedStacks.containsKey(stackId);
+    }
+
+    public void setStack(UUID stackId, Resource<StackEntity> stack) {
+        cachedStacks.put(stackId, stack);
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/util/StacksCache.java:22]: Maybe `putIfAbsent` would be slightly more optimal

File: src/test/java/com/sap/cloudfoundry/client/facade/CloudStackIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+                                          .addBuildpack(IntegrationTestConstants.JAVA_BUILDPACK)
+                                          .build();
+        try {
+            verifyExistenceOfDefaultStack(applicationName, staging, Set.of(DEFAULT_ROUTE_SUMMARY));
+        } finally {
+            client.deleteApplication(applicationName);
+        }
+    }
+
+    private void verifyExistenceOfDefaultStack(String applicationName, Staging staging, Set<CloudRouteSummary> cloudRoutesSummary) {
```
Comment (radito3) [inline src/test/java/com/sap/cloudfoundry/client/facade/CloudStackIntegrationTest.java:64]: These are multiple cloud routes, not a single summary of many routes. Rename the 3rd variable

File: src/test/java/com/sap/cloudfoundry/client/facade/CloudStackIntegrationTest.java [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+                                          .addBuildpack(IntegrationTestConstants.JAVA_BUILDPACK)
+                                          .build();
+        try {
+            verifyIncorrectStack(applicationName, staging, Set.of(DEFAULT_ROUTE_SUMMARY));
+        } finally {
+            client.deleteApplication(applicationName);
+        }
+    }
+
+    private void verifyIncorrectStack(String applicationName, Staging staging, Set<CloudRouteSummary> cloudRoutesSummary) {
```
Comment (radito3) [inline src/test/java/com/sap/cloudfoundry/client/facade/CloudStackIntegrationTest.java:103]: Same here

File: src/test/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudApplicationTest.java [hunk 22]
```diff
@@ -6,9 +6,11 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.UUID;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
+import com.sap.cloudfoundry.client.facade.domain.CloudMetadata;
```
Comment (radito3) [inline src/test/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudApplicationTest.java:13]: Unused imports UUID & CloudMetadata

File: src/main/java/com/sap/cloudfoundry/client/facade/util/StacksCache.java [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+import org.cloudfoundry.client.v2.stacks.StackEntity;
+import org.springframework.util.ConcurrentReferenceHashMap;
+
+import java.util.UUID;
+
+public class StacksCache {
+
+    private final ConcurrentReferenceHashMap<UUID, Resource<StackEntity>> cachedStacks = new ConcurrentReferenceHashMap<>();
+
+    public Resource<StackEntity> getCloudStack(UUID stackId) {
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/util/StacksCache.java:13]: Rename this method

File: src/test/java/com/sap/cloudfoundry/client/facade/CloudStackIntegrationTest.java [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+        client.createApplication(app.getName(), staging, DISK_IN_MB, MEMORY_IN_MB, routeSummaries);
+        assertThrows(CloudOperationException.class, () -> client.getStack(INCORRECT_STACK_NAME, true));
+    }
+
+    @Test
+    @DisplayName("Verify existance of at least one valid Cloud Stack")
+    void getStackList() {
+        assertTrue(client.getStacks() != null);
+        assertTrue(client.getStacks()
+                         .size() > 0);
```
Comment (radito3) [inline src/test/java/com/sap/cloudfoundry/client/facade/CloudStackIntegrationTest.java:120]: `assertTrue(!client.getStacks().isEmpty())` _or_ `assertFalse(client.getStacks().isEmpty())`

**Review by radito3 — APPROVED** [review]
